src/digit/mod.rs

```diff
@@ -56,27 +57,311 @@ pub trait Digit: AsRef<str> {
+
+        let mut final_value: N = num_traits::Zero::zero();
+        let mut intermediary_value: N = num_traits::Zero::zero();
+
+        let check_add = |this: N, that: N| this.checked_add(&that).ok_or(CANT_CONVERT);
+
+        let checked_mul = |this: N, that: N| this.checked_mul(&that).ok_or(CANT_CONVERT);
+
+        for t in parsed {
+            match t {
+                TokenType::FaceValue(v) => {
+                    let v_n: N = v.try_into().map_err(|_| CANT_CONVERT)?;
+
+                    intermediary_value = check_add(intermediary_value, v_n)?;
+                    last = Some(TokenType::FaceValue(v));
+                }
+                TokenType::Multiplier(m) => {
+                    if last.map_or(false, |last| matches!(last, TokenType::Multiplier(_))) {
+                        return Err("Incorrect format: two multipliers in a row");
+                    }
+
+                    // a bit of helper: if this is the first iteration, you can omit a 'یک' and we
+                    // replace it here.
+                    if last.is_none() {
+                        intermediary_value = num_traits::One::one();
+                    }
+
+                    let m_n: N = m.try_into().map_err(|_| CANT_CONVERT)?;
+                    intermediary_value = checked_mul(intermediary_value, m_n)?;
+                    final_value = check_add(final_value, intermediary_value)?;
+
+                    intermediary_value = num_traits::Zero::zero();
+                    last = Some(TokenType::Multiplier(m));
+                }
+            }
+        }
+
+        if let Some(TokenType::FaceValue(_)) = last {
+            final_value = final_value
+                .checked_add(&intermediary_value)
+                .ok_or(CANT_CONVERT)?;
+        }
+
+        Ok(final_value)
+    }
+}
+
+impl WordsToNumber for String {}
+impl WordsToNumber for str {}
+
+#[cfg(test)]
+mod word_to_number {
+    use super::*;
+
+    #[test]
+    fn words_to_number_basic() {
+        // a random example to begin with.
+        assert_eq!(
+            "هفتصد و بیست و یک هزار و دویست و بیست و یک"
+                .words_to_number::<u32>()
+                .unwrap(),
+            721221,
+        );
+    }
+
+    #[test]
+    fn face_value_only() {
+        // face value only.
+        assert_eq!("یک".words_to_number::<u32>().unwrap(), 1);
+        assert_eq!("یازده".words_to_number::<u32>().unwrap(), 11);
+    }
+
+    #[test]
+    fn one_multiplier() {
+        // 1 multiplier
+        assert_eq!("بیست و یک".words_to_number::<u32>().unwrap(), 21);
+        assert_eq!("نود و نه".words_to_number::<u32>().unwrap(), 99);
+        assert_eq!("دویست و هشت".words_to_number::<u32>().unwrap(), 208);
+
+        assert_eq!("هزار و یازده".words_to_number::<u32>().unwrap(), 1011);
+        assert_eq!("یک هزار و یازده".words_to_number::<u32>().unwrap(), 1011);
+
+        assert_eq!(
+            "میلیون و یازده".words_to_number::<u32>().unwrap(),
```

Comment:
Persian is buggy here, I never thought about this... I think, "Million o yaazdah" is wrong, but "Hezar o yaazdah" is correct 🤔 

I think we can leave this bug to the language, not our code :D